timeseries/src/autogluon/timeseries/splitter.py

```diff
@@ -135,54 +129,70 @@ class MultiWindowSplitter(AbstractTimeSeriesSplitter):
-        if num_too_short_series == ts_dataframe.num_items:
-            raise ValueError(f"{self.name} produced an empty training set since all sequences are too short")
-
-        validation_dataframes = [append_suffix_to_item_id(ts_dataframe, "_[None:None]")]
-
-        for window_idx in range(1, self.num_windows):
-            ts_dataframe = ts_dataframe.slice_by_timestep(slice(None, -step_size))
-            total_offset = step_size * window_idx
-            next_val_dataframe = append_suffix_to_item_id(ts_dataframe, f"_[None:{-total_offset}]")
-            validation_dataframes.append(next_val_dataframe)
-
-        train_data = ts_dataframe.slice_by_timestep(slice(None, -prediction_length))
+        original_item_order = ts_dataframe.item_ids
+        original_freq = ts_dataframe.freq
+
+        train_dataframes = []
+        validation_dataframes = []
+        for window_idx in range(self.num_windows):
+            num_timesteps_per_item = ts_dataframe.num_timesteps_per_item()
+
+            item_index = num_timesteps_per_item.index
+            long_enough = num_timesteps_per_item > 2 * prediction_length
+            # Convert boolean indicator into item_id index
+            can_be_split = item_index[long_enough]
+            cannot_be_split = item_index[~long_enough]
+
+            train_dataframes.append(ts_dataframe.loc[cannot_be_split])
+            # Keep timeseries that are long enough for the next round of splitting
+            ts_dataframe = ts_dataframe.loc[can_be_split]
+
+            if window_idx == 0:
+                suffix = "_[None:None]"
+                # TODO: Should we also warn users if there are too few items in the validation set?
+                if len(can_be_split) == 0:
+                    raise ValueError(
+                        f"Cannot create a validation set because all training time series are too short. "
+                        f"At least some time series in train_data must have length >= 2 * prediction_length + 1 "
+                        f"(at least {2 * prediction_length + 1}) but the longest training series has length "
+                        f"{num_timesteps_per_item.max()}. Please decrease prediction_length, provide longer "
+                        f"time series in train_data, or provide tuning_data."
+                    )
+            else:
+                suffix = f"_[None:-{window_idx * prediction_length}]"
+                if len(can_be_split) == 0:
+                    break
+
+            validation_dataframes.append(append_suffix_to_item_id(ts_dataframe, suffix))
+            ts_dataframe = ts_dataframe.slice_by_timestep(slice(None, -prediction_length))
+
+        train_dataframes.append(ts_dataframe)
+
+        train_data = pd.concat(train_dataframes).loc[original_item_order]
+        train_data._cached_freq = original_freq
         val_data = pd.concat(validation_dataframes)
-        val_data._cached_freq = train_data._cached_freq
```

Comment:
out of curiosity, do the freqs change, so slice does not preserve cached freqs? and why change this line specifically?

Reply:
`_cached_freq` isn't preserved during `concat`. We could keep it as-is, but this makes it more transparent that we just use the `freq` of the original dataset as `_cached_freq` for both `train_data` and `val_data`.